### PR TITLE
add `toSignedInt`, closes #72

### LINKED ISSUE #72: Type conversion for small BigInts into integers
Since BigInt's limb parameter has become private, thus we can not easily cast BigInts with absolute value less than `2^32` to an `int` or even a `uint32`. 
The only workaround yet is to use the keyword `cast[int]`. But this does not always do what we want.
To get only the smallest limb, we also have to define a mask like `2^32 - 1` and to apply an ineffective `and`.
Can we implement some functions like … ?:
- `toUint32(a: BigInt): uint32` returns the only limb for BigInts between 1-2^32 and +2^32-1 (small BigInts).
-  `toInt(a: BigInt): int` cast the BigInt if it is small enough, into an int, according to the sign of BigInt, with correct representation (2's complement) and raises a ValueError if the cast can not be done (if int is stored on 32 bits, it may not fit if the heavy weight bit of the uint32 corresponding to first limb is 1).
- `toUint64(a: BigInt): uint64` returns the uint64 made up of one to two limbs, if the BigInt is small enough.
-  `toFloat` like described in the similar issue #34.

### PR DESCRIPTION
**Pull request by narimiran**

### CODE REVIEW

File: src/bigints.nim [hunk 1]
```diff
@@ -753,6 +753,45 @@ func `divmod`*(a, b: BigInt): tuple[q, r: BigInt] =
...
+      result = some(T(x.limbs[0]))
+    if result.isSome and x.isNegative:
+      if result.get == int.low: return # a hack to prevent underflow
+      result = some(not(result.get - 1))
+  else:
+    if x.limbs.len > 1:
+      result = none(T)
+    else:
+      if x.isNegative:
+        if x.limbs[0] - 1 > uint32(T.high):
```
Comment (konsumlamm) [inline src/bigints.nim:784]: What if `x.limbs[0]` is 0? That's not disallowed currently afaik.

**Comment by konsumlamm** [comment]
Nice! I think we should also support unsigned integers, is there any particular reason you didn't include them here? `toInt` would probably be a better name then, unless we want a separate `toUnsignedInt` function (which would be kinda weird since we already have to specify the int type anyway)?

File: tests/tbigints.nim [hunk 2]
```diff
@@ -384,5 +384,42 @@ template main() =
...
     doAssertRaises(DivByZeroDefect): discard divmod(one, zero)
 
+  block: # toSignedInt
+    let
+      a = initBigInt(7)
+      b = initBigInt(-7)
+      c = initBigInt(int32.high)
+      d = initBigInt(int32.low)
+      e = "123456789123456789123456789".initBigInt
+      f = "340282366920938463463374607431768211456".initBigInt
```
Comment (dlesnoff) [inline tests/tbigints.nim:394]: Could you add a comment to tell how many limbs are necessary for them to be represented ?
I guess log_2(e) = 87 bits so e needs 3 limbs and 128 bits for f(=2^128) that sums to 4 limbs.

File: src/bigints.nim [hunk 3]
```diff
@@ -753,6 +753,45 @@ func `divmod`*(a, b: BigInt): tuple[q, r: BigInt] =
...
+      b = 130.initBigInt
+    assert toSignedInt[int8](a) == some(44'i8)
+    assert toSignedInt[int8](b) == none(int8)
+    assert toSignedInt[int](b) == some(130)
+
+  when T is int64 or T is int:
+    if x.limbs.len > 2:
+      result = none(T)
+    elif x.limbs.len == 2:
+      result = some((T(x.limbs[1]) shl 32) + T(x.limbs[0]))
```
Comment (dlesnoff) [inline src/bigints.nim:773]: What happens if you try to convert a number that is between int64.high (=2^63-1) and 2^64 ?

File: src/bigints.nim [hunk 4]
```diff
@@ -753,6 +753,45 @@ func `divmod`*(a, b: BigInt): tuple[q, r: BigInt] =
...
+  runnableExamples:
+    import options
+    let
+      a = 44.initBigInt
+      b = 130.initBigInt
+    assert toSignedInt[int8](a) == some(44'i8)
+    assert toSignedInt[int8](b) == none(int8)
+    assert toSignedInt[int](b) == some(130)
+
+  when T is int64 or T is int:
```
Comment (dlesnoff) [inline src/bigints.nim:769]: As I understand the type, `int` uses 32 bit on a 32 bit architecture and 64 bits on a 64 bits architecture (or depending on the compiler’s flags ?). The code under the `when` seems to assume T uses 64 bit and I fear that might not always be the case if `T` is an `int`.

File: tests/tbigints.nim [hunk 5]
```diff
@@ -384,5 +384,42 @@ template main() =
...
+      g = initBigInt(int64.high)
+      h = initBigInt(int64.low)
+      i = initBigInt(int64.low + 1)
```
Comment (dlesnoff) [inline tests/tbigints.nim:397]: Could we test for int64.high + 1 and int64.low - 1 ?

Comment (konsumlamm) [reply tests/tbigints.nim:397]: That would be over-/underflow, what would you expect the test to do?

File: src/bigints.nim (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlesnoff) [reply src/bigints.nim:769]: Why > 4 ? Can we just not test equality `sizeof(T) == 8` ? (Unless we target 128-bit architecture)

File: tests/tbigints.nim (hunk 5 above)
Comment (dlesnoff) [reply tests/tbigints.nim:397]: To return a none Option. But as of right now, I believe the code will make an over-/underflow.

File: src/bigints.nim (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlesnoff) [reply src/bigints.nim:769]: As `int` is architecture dependent, the code under the when seems to assume T uses 64 bit and I fear that might not always be the case if T is an int and architecture is 32-bit.

File: src/bigints.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (dlesnoff) [reply src/bigints.nim:773]: I believe this does not work for integers between int64.high (=2^63-1) and 2^64.

File: tests/tbigints.nim (hunk 5 above)
Comment (konsumlamm) [reply tests/tbigints.nim:397]: `int64.high + 1` and `int64.low - 1` can't be represented in 64 bits, so I don't know how you expect this to work. I'm currently not sure what the expected behaviour for signed over-/underflow is (wraparound, exception, UB), but either way, it won't be the same as `initBigInt(int64.high) + initBigInt(1)`/`initBigInt(int64.low) - initBigInt(1)`.

File: src/bigints.nim (hunk 4 above)
[the remark this answers is not in the document]
Comment (narimiran) [reply src/bigints.nim:769]: Yeah, testing for `sizeof` is definitely better.

**Comment by narimiran** [comment]
> I think we should also support unsigned integers, is there any particular reason you didn't include them here? 

To get your comments, improve the code, and then based on that, continue with unsigned.

File: tests/tbigints.nim (hunk 5 above)
Comment (narimiran) [reply tests/tbigints.nim:397]: > Could we test for int64.high + 1 and int64.low - 1 ?

We can.

> what would you expect the test to do?

To return `none`.

File: src/bigints.nim (hunk 1 above)
Comment (narimiran) [reply src/bigints.nim:784]: Can we have a negative zero? *Should* we have a negative zero?

File: src/bigints.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (narimiran) [reply src/bigints.nim:773]: > e.g. `0xffffffff40000000` gives the wrong result

Now that's a good candidate to include in the tests :)

File: src/bigints.nim (hunk 1 above)
Comment (konsumlamm) [reply src/bigints.nim:784]: Currently yes, for example `-0.initBigInt` (it gets treated the same as any other zero though). Wether that *should* be allowed is a different question.

File: src/bigints.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (narimiran) [reply src/bigints.nim]: See the updated code.

File: src/bigints.nim [hunk 6]
```diff
@@ -753,6 +753,53 @@ func `divmod`*(a, b: BigInt): tuple[q, r: BigInt] =
...
+    assert toSignedInt[int8](b) == none(int8)
+    assert toSignedInt[int](b) == some(130)
+
+  when sizeof(T) == 8:
+    if x.limbs.len > 2:
+      result = none(T)
+    elif x.limbs.len == 2:
+      if x.limbs[1] > uint32.high shr 1:
+        if x.isNegative and x.limbs[0] == 0:
+          result = some(T(int64.low))
```
Comment (konsumlamm) [inline src/bigints.nim:775]: This breaks for e.g. `-initBigInt(0xffffffff00000000'u64)`, you also need to check `x.limbs[1]`.

**Review by konsumlamm — COMMENTED** [review]
I think it would be easier/cleaner to fully convert the absolute values to unsigned ints (the only checks required for that are checking the length of `x.limbs`) and then convert that to signed ints.

**Comment by Araq** [comment]
I need these additions so I'm merging it. Later on more improvements can follow. I also don't mind to change my code if you decide to tune the API.

**Pull request merged**